### PR TITLE
CSP-419 POST Admin member lookup

### PR DESCRIPTION
**Pull request by MarkFCain**

### CODE REVIEW

File: src/ApprenticeAan/SFA.DAS.ApprenticeAan.Application/Admins/Queries/Lookup/LookupAdminMemberRequest.cs [hunk 1]
```diff
@@ -0,0 +1,9 @@
+﻿using MediatR;
+
+namespace SFA.DAS.ApprenticeAan.Application.Admins.Queries.Lookup;
+public class LookupAdminMemberRequest : IRequest<LookupAdminMemberResult?>
+{
+    public string Email { get; set; } = null!;
+    public string FirstName { get; set; } = null!;
+    public string LastName { get; set; } = null!;
```
Comment (DevenKShah) [inline src/ApprenticeAan/SFA.DAS.ApprenticeAan.Application/Admins/Queries/Lookup/LookupAdminMemberRequest.cs:8]: These fields are not required by the handler. I see why they are in here, I guess you need to use a model to take the incoming request body and then pass only email here.

File: src/ApprenticeAan/SFA.DAS.ApprenticeAan.Application/Admins/Queries/Lookup/LookupAdminMemberHandler.cs [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+    {
+        _apiClient = apiClient;
+    }
+
+    public async Task<LookupAdminMemberResult?> Handle(LookupAdminMemberRequest request,
+        CancellationToken cancellationToken)
+    {
+        var result = await _apiClient.GetMemberByEmail(request.Email, cancellationToken);
+
+        return result.ResponseMessage.StatusCode != System.Net.HttpStatusCode.OK
```
Comment (DevenKShah) [inline src/ApprenticeAan/SFA.DAS.ApprenticeAan.Application/Admins/Queries/Lookup/LookupAdminMemberHandler.cs:20]: Invert to use positive check `if == Ok ? return result : null`

File: src/ApprenticeAan/SFA.DAS.ApprenticeAan.Api/Controllers/AdminsController.cs [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+    [HttpPost]
+    [ProducesResponseType(typeof(LookupAdminMemberResult), StatusCodes.Status200OK)]
+    public async Task<IActionResult> Lookup([FromBody] LookupAdminMemberRequest request, CancellationToken cancellationToken)
+    {
+        var result = await _mediator.Send(request, cancellationToken);
+
+        if (result != null)
+            return Ok(result);
+
+        var createAdminResult = await _mediator.Send(new CreateAdminMemberCommand { Email = request.Email, FirstName = request.FirstName, LastName = request.LastName }, cancellationToken);
```
Comment (DevenKShah) [inline src/ApprenticeAan/SFA.DAS.ApprenticeAan.Api/Controllers/AdminsController.cs:28]: Could have used operator here for conversion.

File: src/ApprenticeAan/SFA.DAS.ApprenticeAan.Api/Controllers/AdminsController.cs [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+    }
+
+    [HttpPost]
+    [ProducesResponseType(typeof(LookupAdminMemberResult), StatusCodes.Status200OK)]
+    public async Task<IActionResult> Lookup([FromBody] LookupAdminMemberRequest request, CancellationToken cancellationToken)
+    {
+        var result = await _mediator.Send(request, cancellationToken);
+
+        if (result != null)
+            return Ok(result);
```
Comment (DevenKShah) [inline src/ApprenticeAan/SFA.DAS.ApprenticeAan.Api/Controllers/AdminsController.cs:26]: `if` with no block should be in single line

File: src/ApprenticeAan/SFA.DAS.ApprenticeAan.Api.UnitTests/Controllers/AdminsControllerTests.cs [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+    {
+        mediatorMock.Setup(m => m.Send(request, cancellationToken)).ReturnsAsync(expected);
+
+        var response = await sut.Lookup(request, cancellationToken);
+
+        mediatorMock.Verify(m => m.Send(request, cancellationToken));
+        response.As<OkObjectResult>().Value.Should().Be(expected);
+    }
+
+    [Test, MoqAutoData]
```
Comment (DevenKShah) [inline src/ApprenticeAan/SFA.DAS.ApprenticeAan.Api.UnitTests/Controllers/AdminsControllerTests.cs:32]: LookupMember_MemberDoesNotExist_InvokesCreateAdminMember

File: src/ApprenticeAan/SFA.DAS.ApprenticeAan.Api.UnitTests/Controllers/AdminsControllerTests.cs [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+using SFA.DAS.ApprenticeAan.Application.Admins.Queries.Lookup;
+using SFA.DAS.ApprenticeAan.Application.Infrastructure.Configuration;
+using SFA.DAS.Testing.AutoFixture;
+
+namespace SFA.DAS.ApprenticeAan.Api.UnitTests.Controllers;
+
+public class AdminsControllerTests
+{
+    [Test, MoqAutoData]
+    public async Task LookupMemberWithDetailPresent_InvokesRequest(
```
Comment (DevenKShah) [inline src/ApprenticeAan/SFA.DAS.ApprenticeAan.Api.UnitTests/Controllers/AdminsControllerTests.cs:17]: LookupMember_MemberExists_InvokesRequest

**Review by DevenKShah — DISMISSED** [review]

**Review by DevenKShah — COMMENTED** [review]
Looks good, just need to shuffle files around.

File: src/AdminAan/SFA.DAS.AdminAan/Application/Admins/Queries/Lookup/GetMemberResult.cs
[the remark this answers is not in the document]
Comment (MarkFCain) [reply src/AdminAan/SFA.DAS.AdminAan/Application/Admins/Queries/Lookup/GetMemberResult.cs:1]: ok, have changed, but I did forsee that having the GetMemberByEmail complete results might be useful in other areas.

File: src/AdminAan/SFA.DAS.AdminAan.UnitTests/SFA.DAS.AdminAan.UnitTests.csproj [hunk 7]
```diff
@@ -1,4 +1,4 @@
-<Project Sdk="Microsoft.NET.Sdk">
+﻿<Project Sdk="Microsoft.NET.Sdk">
```
[the remark this answers is not in the document]
Comment (MarkFCain) [reply src/AdminAan/SFA.DAS.AdminAan.UnitTests/SFA.DAS.AdminAan.UnitTests.csproj:1]: 🤷

File: src/AdminAan/SFA.DAS.AdminAan.Api/Controllers/AdminsController.cs [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+    private readonly IMediator _mediator;
+
+    public AdminsController(IMediator mediator)
+    {
+        _mediator = mediator;
+    }
+
+    [HttpPost]
+    [ProducesResponseType(typeof(LookupAdminMemberResult), StatusCodes.Status200OK)]
+    public async Task<IActionResult> Lookup([FromBody] LookupAdminMemberRequest request, CancellationToken cancellationToken)
```
[the remark this answers is not in the document]
Comment (MarkFCain) [reply src/AdminAan/SFA.DAS.AdminAan.Api/Controllers/AdminsController.cs]: ok done

**Review by DevenKShah — APPROVED** [review]

**Pull request merged**